File: mobile/android/PairMobileAndroid/src/main/kotlin/dev/pair/mobile/android/PairAuthSessionManager.kt
package dev.pair.mobile.android

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Bootstrap and token result for managed mobile authentication sessions. */
sealed class PairAuthSessionManagerResult<out Session> {

    /** The saved session was refreshed when needed and validated. */
    data class Valid<Session>(val session: Session) : PairAuthSessionManagerResult<Session>()

    /** No saved session exists. */
    data object Missing : PairAuthSessionManagerResult<Nothing>()

    /** The saved session is available locally, but remote validation could not complete. */
    data class Offline<Session>(val session: Session) : PairAuthSessionManagerResult<Session>()

    /** The backend rejected the saved session. */
    data object Invalidated : PairAuthSessionManagerResult<Nothing>()
}

/** Access-token result for callers that need an authorization header value. */
sealed class PairAccessTokenResult<out Session> {

    /** A usable access token is available. */
    data class Valid<Session>(val accessToken: String, val session: Session) : PairAccessTokenResult<Session>()

    /** No saved session exists. */
    data object Missing : PairAccessTokenResult<Nothing>()

    /** The saved session is available locally, but refresh could not complete. */
    data class Offline<Session>(val session: Session) : PairAccessTokenResult<Session>()

    /** The backend rejected the saved session. */
    data object Invalidated : PairAccessTokenResult<Nothing>()
}

/** Coordinates persistent mobile sessions, startup validation, and single-flight token refresh. */
class PairAuthSessionManager<User>(
    private val store: PairSessionStore<PairStoredAuthSession<User>>,
    private val refreshLeewaySeconds: Long = 60,
    private val nowEpochSeconds: () -> Long = PairClock::epochSeconds
) {
    private val refreshMutex = Mutex()

    /** Reads the persisted session snapshot. */
    suspend fun load(): PairStoredAuthSession<User>? = store.load()

    /** Saves a newly authenticated or externally migrated session snapshot. */
    suspend fun save(session: PairStoredAuthSession<User>) {
        store.save(session)
    }

    /** Clears the persisted session snapshot. */
    suspend fun clear() {
        store.clear()
    }

    /** Restores a session, refreshes stale tokens, and validates the result before app-private UI is shown. */
    suspend fun bootstrap(
        validate: suspend (PairStoredAuthSession<User>) -> PairStoredAuthSession<User>,
        refresh: suspend (PairStoredAuthSession<User>) -> PairStoredAuthSession<User>
    ): PairAuthSessionManagerResult<PairStoredAuthSession<User>> {
        val session = store.load() ?: return PairAuthSessionManagerResult.Missing

        return try {
            val tokenSession = sessionWithValidAccessToken(session, refresh)
            val validatedSession = validate(tokenSession)
            store.save(validatedSession)
            PairAuthSessionManagerResult.Valid(validatedSession)
        } catch (error: Throwable) {
            if (isDefinitiveAuthenticationFailure(error)) {
                store.clear()
                PairAuthSessionManagerResult.Invalidated
            } else {
                PairAuthSessionManagerResult.Offline(store.load() ?: session)
            }
        }
    }

    /** Returns a usable access token, refreshing once for all concurrent callers when needed. */
    suspend fun validAccessToken(
        refresh: suspend (PairStoredAuthSession<User>) -> PairStoredAuthSession<User>
    ): PairAccessTokenResult<PairStoredAuthSession<User>> {
        val session = store.load() ?: return PairAccessTokenResult.Missing

        return try {
            val tokenSession = sessionWithValidAccessToken(session, refresh)
            PairAccessTokenResult.Valid(tokenSession.accessToken, tokenSession)
        } catch (error: Throwable) {
            if (isDefinitiveAuthenticationFailure(error)) {
                store.clear()
                PairAccessTokenResult.Invalidated
            } else {
                PairAccessTokenResult.Offline(session)
            }
        }
    }

    /** Returns the input session when its token is usable or refreshes it through the shared lock. */
    private suspend fun sessionWithValidAccessToken(
        session: PairStoredAuthSession<User>,
        refresh: suspend (PairStoredAuthSession<User>) -> PairStoredAuthSession<User>
    ): PairStoredAuthSession<User> {
        if (session.hasUsableAccessToken(nowEpochSeconds(), refreshLeewaySeconds)) {
            return session
        }

        if (session.refreshToken.isNullOrBlank()) {
            throw MissingRefreshTokenException
        }

        return singleFlightRefresh(refresh)
    }

    /** Runs one refresh operation while concurrent callers wait and reuse the stored result. */
    private suspend fun singleFlightRefresh(
        refresh: suspend (PairStoredAuthSession<User>) -> PairStoredAuthSession<User>
    ): PairStoredAuthSession<User> = refreshMutex.withLock {
        val currentSession = store.load() ?: throw MissingRefreshTokenException

        if (currentSession.hasUsableAccessToken(nowEpochSeconds(), refreshLeewaySeconds)) {
            return@withLock currentSession
        }

        if (currentSession.refreshToken.isNullOrBlank()) {
            throw MissingRefreshTokenException
        }

        val refreshedSession = refresh(currentSession)
        store.save(refreshedSession)

        refreshedSession
    }

    /** Classifies only explicit auth failures as reasons to remove a local snapshot. */
    private fun isDefinitiveAuthenticationFailure(error: Throwable): Boolean =
        error is MissingRefreshTokenException || (error is PairApiException && error.isAuthenticationFailure)
}

private data object MissingRefreshTokenException : Exception("Missing refresh token.")

File: mobile/ios/PairMobileKit/Sources/PairMobileKit/PairAuthSessionManager.swift
import Foundation

/// Bootstrap and token result for managed mobile authentication sessions.
public enum PairAuthSessionManagerResult<Session: Sendable>: Sendable {
	case valid(Session)
	case missing
	case offline(Session)
	case invalidated
}

/// Access-token result for callers that need an authorization header value.
public enum PairAccessTokenResult<Session: Sendable>: Sendable {
	case valid(accessToken: String, session: Session)
	case missing
	case offline(Session)
	case invalidated
}

/// Coordinates persistent mobile sessions, startup validation, and single-flight token refresh.
public actor PairAuthSessionManager<User: Codable & Sendable, Context: Codable & Sendable> {
	public typealias Session = PairStoredAuthSession<User, Context>

	private let store: PairKeychainStore<Session>
	private let refreshLeeway: TimeInterval
	private let now: @Sendable () -> Date
	private var refreshTask: Task<Session, Error>?

	/// Initializes the manager with a Keychain store and the pre-expiration refresh window.
	public init(
		store: PairKeychainStore<Session>,
		refreshLeeway: TimeInterval = 60,
		now: @escaping @Sendable () -> Date = Date.init
	) {
		self.store = store
		self.refreshLeeway = max(refreshLeeway, 0)
		self.now = now
	}

	/// Reads the persisted session snapshot.
	public func load() async -> Session? {
		await store.load()
	}

	/// Saves a newly authenticated or externally migrated session snapshot.
	public func save(_ session: Session) async {
		await store.save(session)
	}

	/// Clears the persisted session snapshot.
	public func clear() async {
		await store.clear()
	}

	/// Restores a session, refreshes stale tokens, and validates the result before app-private UI is shown.
	public func bootstrap(
		validate: @Sendable @escaping (Session) async throws -> Session,
		refresh: @Sendable @escaping (Session) async throws -> Session
	) async -> PairAuthSessionManagerResult<Session> {
		guard let session = await store.load() else {
			return .missing
		}

		do {
			let tokenSession = try await sessionWithValidAccessToken(session, refresh: refresh)
			let validatedSession = try await validate(tokenSession)

			await store.save(validatedSession)

			return .valid(validatedSession)
		} catch {
			if isDefinitiveAuthenticationFailure(error) {
				await store.clear()
				return .invalidated
			}

			return .offline(await store.load() ?? session)
		}
	}

	/// Returns a usable access token, refreshing once for all concurrent callers when needed.
	public func validAccessToken(
		refresh: @Sendable @escaping (Session) async throws -> Session
	) async -> PairAccessTokenResult<Session> {
		guard let session = await store.load() else {
			return .missing
		}

		do {
			let tokenSession = try await sessionWithValidAccessToken(session, refresh: refresh)

			return .valid(accessToken: tokenSession.accessToken, session: tokenSession)
		} catch {
			if isDefinitiveAuthenticationFailure(error) {
				await store.clear()
				return .invalidated
			}

			return .offline(session)
		}
	}

	/// Forces a token refresh and shares that refresh with concurrent callers.
	public func refreshAccessToken(
		refresh: @Sendable @escaping (Session) async throws -> Session
	) async -> PairAccessTokenResult<Session> {
		guard let session = await store.load() else {
			return .missing
		}

		do {
			let refreshedSession = try await singleFlightRefresh(session, refresh: refresh)

			return .valid(accessToken: refreshedSession.accessToken, session: refreshedSession)
		} catch {
			if isDefinitiveAuthenticationFailure(error) {
				await store.clear()
				return .invalidated
			}

			return .offline(session)
		}
	}

	/// Returns the input session when its token is usable or refreshes it through the shared task.
	private func sessionWithValidAccessToken(
		_ session: Session,
		refresh: @Sendable @escaping (Session) async throws -> Session
	) async throws -> Session {
		if isAccessTokenUsable(session) {
			return session
		}

		guard session.refreshToken != nil else {
			throw PairAuthSessionManagerInternalError.missingRefreshToken
		}

		return try await singleFlightRefresh(session, refresh: refresh)
	}

	/// Runs one refresh operation and lets later callers await the same result.
	private func singleFlightRefresh(
		_ session: Session,
		refresh: @Sendable @escaping (Session) async throws -> Session
	) async throws -> Session {
		if let refreshTask {
			return try await refreshTask.value
		}

		let task = Task<Session, Error> {
			try await refresh(session)
		}

		refreshTask = task

		do {
			let refreshedSession = try await task.value

			await store.save(refreshedSession)
			refreshTask = nil

			return refreshedSession
		} catch {
			refreshTask = nil
			throw error
		}
	}

	/// Checks whether the access token remains outside the pre-expiration refresh window.
	private func isAccessTokenUsable(_ session: Session) -> Bool {
		session.expiresAt.timeIntervalSince(now()) > refreshLeeway
	}

	/// Classifies only explicit auth failures as reasons to remove a local snapshot.
	private func isDefinitiveAuthenticationFailure(_ error: Error) -> Bool {
		if let apiError = error as? PairAPIError {
			return apiError.isAuthenticationFailure
		}

		if case PairAuthSessionManagerInternalError.missingRefreshToken = error {
			return true
		}

		return false
	}
}

private enum PairAuthSessionManagerInternalError: Error {
	case missingRefreshToken
}
